### ISSUE TITLE
screen passes incorrect argument to Function callable

### ISSUE THREAD

**Issue by Numerlor** [renpy/renpy#3718]
When running renpy 8 with the following script code
```renpy
init python:
    def f(o):
        print(f"running function {id(o)=}")

screen Test():
    default obj = []
    textbutton "dummy":
        align (0.5, 0.5)
        action Function(f, obj)
    $ print(f"running screen {id(obj)=}\n")


label start:
    show screen Test
    "1"
```
renpy passes a wrong list to the function called by the action. Running the game and clicking the button twice, the following output is produced (concrete values of IDs irrelevant):
```
running screen id(obj)=86317312  # in main menu

# game started
running screen id(obj)=86296752  

running screen id(obj)=86296752

# first click
running function id(o)=86317312
running screen id(obj)=86296752

# second click
running function id(o)=86317312
running screen id(obj)=86296752
```
You can see that the object from the first run of the screen is passed to the function, while the screen is using a new one.
This issue was present when I tried it with renpy's custom lists, dicts, and set, but not when done with a plain `object` instance or an user defined class. 

The error does also seem to correct itself on the second run of the action if the value of the object is changed (e.g. appended to the list), but the first run of the action is a noop from the screen's POV as the function acts on an object different to the one used by the screen.

**Comment by Gouvernathor** [comment]
Python code [included in screens](https://www.renpy.org/doc/html/screens.html#python) is not supposed to cause side effects - and printing to the console is, by definition, a side-effect.
Do you have an example of this causing a problematic behavior in a practical use-case ?

**Comment by Numerlor** [comment]
This prints are there only for the sake of showing what's happening, something like this showcases the behaviour in a user facing manner
```renpy
init python:
    def f(o):
        o.append(0)

screen Test():
    default obj = []
    textbutton "dummy":
        align (0.5, 0.5)
        action Function(f, obj)
    if obj:
        text "this should show up after first click":
            align (0.5, 0.5)
            yoffset 20
```

**Comment by via-my** [comment]
You can get variables from screens in a different way. Here's my workaround I use.

```
def set_screen_variable(variable, value, screen=None):
    """
    Set value to the variable of the current or specified screen.
    """
    if screen is None:
        cs = renpy.current_screen()
    else:
        cs = renpy.get_screen(screen)

    rv = cs.scope[variable] = value


def get_screen_variable(variable, screen=None):
    """
    Get value from the variable of the current or specified screen.
    """
    if screen is None:
        cs = renpy.current_screen()
    else:
        cs = renpy.get_screen(screen)

    if cs is not None:
        return cs.scope.get(variable, None)

    else:
        return None
```

**Comment by Gouvernathor** [comment]
@via-my the `scope` parameter is [not documented](https://www.renpy.org/dev-doc/html/screen_python.html#renpy.get_screen) and should no be used, because as usual a future update may break your code.
What's described here is a bug, as it seems, so It should get solved within renpy.

**Comment by renpytom** [comment]
I still don't see a problem here. When I correct your code to: 

```renpy
init python:
    def f(o):
        o.append(0)
        renpy.restart_interaction()

screen Test():
    default obj = []
    textbutton "dummy":
        align (0.5, 0.5)
        action Function(f, obj)
    if obj:
        text "this should show up after first click":
            align (0.5, 0.5)
            yoffset 20
```

It works. Ren'Py will create additional instances of a screen for prediction purposes, and it looks like that's what's going on. Without the call to renpy.restart_interaction(), there's nothing telling Ren'Py to update the screen, and so we're just seeing prediction activity. 

I'm not convinced there's an issue with Ren'Py here.

**Comment by Numerlor** [comment]
Isn't restart_interaction already being called by the Function action?

**Comment by renpytom** [comment]
You're right, but weirdly, I can't repeat this problem here. Everytime I run your example, I'm getting the correct behavior.

**Comment by Numerlor** [comment]
Tried a fresh nightly build and project, and getting the behaviour from the issue where only the second button press triggers it. fwiw platform.platform() is `Windows-10-10.0.19043-SP0`

**Comment by Andykl** [comment]
I think we have found the source of the problem. The problem will not occur if used code that shows the screen immediately, the prediction needs to do the wrong `Function` evaluation first.

**Comment by Gouvernathor** [comment]
Fixed by 866bb6ce45ee55fb75215d8fd6e75dafd7bc2e8b, fd3e9ce17d3eaf040d7fa9bca494cb95a23f0967 and d961eec4816ea62f36108e0f2d7b5ca690efa411.
